### PR TITLE
feat: [PIE-3591]: More info on expansion for CI with some tests and an example

### PR DESCRIPTION
**Pull request by sahilhindwani-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/35265)

### CODE REVIEW

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 14 of 14 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35265)* status: 14 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyExpansionData.java` line 9 at r1](https://reviewable.io/reviews/harness/harness-core/35265#-N6m6PCv9aCzem6mMP0e:-N6m6PCw8uw_Se6-qsiO:bvqjvfb) ([raw file](https://github.com/harness/harness-core/blob/fdd7097bc8a9e517f962efeb72e60173ba7c8f31/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyExpansionData.java#L9)):*
> ```Java
> @Builder
> public class StrategyExpansionData {
>   int maxConcurrency;
> ```

why this is required then Strategy Info?



_Code quote:_
```Java
public class StrategyExpansionData {
  int maxConcurrency;
```
___
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java` line 123 at r1](https://reviewable.io/reviews/harness/harness-core/35265#-N6m6Ee1CT5to9VYQ8wQ:-N6m6Ee20trdHHva5CaB:bwqzz0y) ([raw file](https://github.com/harness/harness-core/blob/fdd7097bc8a9e517f962efeb72e60173ba7c8f31/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java#L123)):*
> ```Java
>                                           .map(node -> ExecutionWrapperConfig.builder().step(node).uuid(uuid).build())
>                                           .collect(Collectors.toList()))
>             .uuidToStrategyExpansionData(uuidToMaxConcurrency)
> ```

is this code reusable in other if conditions above, then can be extracted as util



_Code quote:_
```Java
        String uuid = EmptyPredicate.isEmpty(executionWrapperConfig.getUuid()) ? UUIDGenerator.generateUuid()
                                                                               : executionWrapperConfig.getUuid();
        uuidToMaxConcurrency.put(
            uuid, StrategyExpansionData.builder().maxConcurrency(strategyInfo.getMaxConcurrency()).build());
        return ExpandedExecutionWrapperInfo.builder()
            .expandedExecutionConfigs(expandedJsonNodes.stream()
                                          .map(node -> ExecutionWrapperConfig.builder().step(node).uuid(uuid).build())
                                          .collect(Collectors.toList()))
            .uuidToStrategyExpansionData(uuidToMaxConcurrency)
```

**Comment by devkimittal** [comment]
It is creating issues when strategy is applied on a step group.  Do not merge this pr.

**Comment by sahilhindwani-harness** [comment]
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java` line 113 at r1](https://reviewable.io/reviews/harness/harness-core/35265#-N6ltqhzF51n297Raxvr-r1-113:-N7E_0iC65i8DStvmuGK:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fdd7097bc8a9e517f962efeb72e60173ba7c8f31/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java#L113)):*
<details><summary><i>Previously, devkimittal (Dev Mittal) wrote…</i></summary><blockquote>

it should be  `expandJsonNodes(executionWrapperConfig.getStepGroup())`
</blockquote></details>

Done.

**Comment by sahilhindwani-harness** [comment]
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java` line 123 at r1](https://reviewable.io/reviews/harness/harness-core/35265#-N6m6Ee1CT5to9VYQ8wQ:-N7EoVng5KGO6NK1wDar:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fdd7097bc8a9e517f962efeb72e60173ba7c8f31/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java#L123)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

is this code reusable in other if conditions above, then can be extracted as util
</blockquote></details>

Done.

**Comment by sahilhindwani-harness** [comment]
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyExpansionData.java` line 9 at r1](https://reviewable.io/reviews/harness/harness-core/35265#-N6m6PCv9aCzem6mMP0e:-N7EoXHk5B2Fx-21Il9J:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fdd7097bc8a9e517f962efeb72e60173ba7c8f31/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyExpansionData.java#L9)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why this is required then Strategy Info?
</blockquote></details>

Done.

**Review by archit-harness — APPROVED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35265)* status: 14 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

File: 878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java [hunk 1]
```diff
@@ -52,44 +59,74 @@ public List<JsonNode> expandJsonNodes(JsonNode nodeWithStrategy) throws IOExcept
...
+          uuidToMaxConcurrency.putAll(expandedExecutionWrapperInfo.getUuidToStrategyExpansionData());
         }
         stepGroupElementConfig.setSteps(executionWrapperConfigs);
         JsonNode stepGroupNode = JsonPipelineUtils.asTree(stepGroupElementConfig);
-        List<JsonNode> expandedJsonNodes = expandJsonNodes(stepGroupNode);
-        return expandedJsonNodes.stream()
-            .map(
-                node -> ExecutionWrapperConfig.builder().stepGroup(node).uuid(executionWrapperConfig.getUuid()).build())
-            .collect(Collectors.toList());
+        StrategyInfo strategyInfo = expandJsonNodes(stepGroupNode);
```
[the remark this answers is not in the document]
Comment (devkimittal) [reply 878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java:113]: I think this is still not changes

**Review by brijeshdhakar — APPROVED** [review]
> Reviewed 11 of 14 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35265)* status: :shipit: complete! ✓

**Pull request merged**